### ISSUE TITLE
Migrate to pcre2

### ISSUE THREAD

**Issue by JanWielemaker** [SWI-Prolog/packages-pcre#2]
[Raised here from Lev Lamberov after E-mail]

I'd like to let you know that old pcre, which is used in swi-prolog,
will be removed from Debian in some time soon [bug-report]. It is
obsolete and upstream stopped to support it (for some historical reasons
it was packaged as pcre3). There is pcre2, which is a modern version of
the library, but typically source code needs some tweaks to work with
it. I don't think simple changes, like include pcre2 headers, will work,
since there are some API-wise differences between pcre and pcre2. You
can find documentation on pcre2 at its upstream web site [pcre2]. It
would be nice if someone (tm) could port swi-prolog to pcre2. Otherwise
when pcre will be removed from Debian (and then from Ubuntu too, I
guess) we will not be able to build swi-prolog with pcre support.

[bug-report] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=999949

[pcre2] https://pcre.org/current/doc/html/

Cheers!

Lev

**Comment by JanWielemaker** [comment]
This issue has been mentioned on **SWI-Prolog**. There might be relevant details there:

https://swi-prolog.discourse.group/t/update-library-regex-to-pcre2-request-for-help/4668/1

**Comment by kamahen** [comment]
I got 400 lines of error messages after making the obvious changes (e.g., `#include <pcre2.h>`) ... most of them look pretty easy to change (e.g., `s/PCRE/PCRE2/`)

One question: what setting should be used for `PCRE2_CODE_UNIT_WIDTH` (see https://stackoverflow.com/questions/46823848/libpcre2-code-unit-width) ... is there documentation on the internal form of strings in swipl?

**Comment by JanWielemaker** [comment]
If I recall well (and a quick check seems to support this), the library is used with UTF-8 encoding.  Prolog's internal representation does not matter.   This is a dual char* and wchar_t* with some plans to eventually move this to UTF-8.

**Comment by kamahen** [comment]
It appears that there are some subtle differences between what otherwise appear to be merely renamings of functions or constants, so I need to do some careful reading of the entire source (1000 lines of C, 600 lines of Prolog). This will take a little while. ;)

And that's on top of larger changes in the API (e.g., prce_study() no longer exists; pcre_exec() is replaced by pcre_match(), etc.)

Style question: the current source has tabs; do you have objections if I replace them by blanks?

**Comment by JanWielemaker** [comment]
> Style question: the current source has tabs; do you have objections if I replace them by blanks?

I prefer to make as little changes as possible.    Tabs to/from spaces can be done automatically though, so it doesn't matter too much.

**Comment by kamahen** [comment]
How about if I convert to blanks (using `expand`) and then put the tabs back using `unexpand --all`? (And do a git squash on my PR, so that the intermediate expanded file won't mess up the "blame")
I did `expand pcre4pl.c | unexpand --all` and result was almost identical (4 lines in the original were missing an internal tab).

**Comment by JanWielemaker** [comment]
Good enough for me :smile:

**Comment by kamahen** [comment]
There are a *lot* of changes to the API in pcre2; I'll try to be backward compatible, but can't be sure that it's 100% (the existing test cases are fairly minimal).

I'm wondering if it might be a good idea to rename the current library(pcre) as library(pcre1) and leave that there for a while, in case someone discovers an incompatible change? When Debian removes PCRE3[0], we would remove library(pcre1).

This should be very little extra work -- basically, just copy the existing files (and leave them in packages-pcre), and mark a few cmake changes as TODOs with a pointer to this bug.

**Comment by JanWielemaker** [comment]
> I'm wondering if it might be a good idea to rename the current library(pcre) as library(pcre1) and leave that there for a while, in case someone discovers an incompatible change? When Debian removes PCRE3[0], we would remove library(pcre1)

I'm not in favor of that.   Two libraries means twice the problems as well as confusion.   People have to explicitly switch to pcre1 while they need to sort possibly application-affecting issues soon anyway.   If anyone has real problems with the new library they can always drop in the old one themselves.  Even Windows users can simply copy the .dll and .pl from an old Prolog version.

The tests are surely incomplete and assumes we only need to tests the interface.   We assume the regex handling of libpcre is correct.

**Comment by kamahen** [comment]
To be clear: the APIs for the pcre1 and pcre2 libraries are very different: you can't replace libpcre.so (or libpcre3.so) libpcre2.so. Upgrading to PCRE2 will require a major rewrite of `pcre4pl.c` and at least some modification of `pcre.pl`.

I think that I can preserve the current API in `pcre.pl`, with the exception of things like `re_config(utf8(X))`, where `PCRE_CONFIG_UTF8` doesn't exist in pcre2, so I'd always return `true` (the pcre1 value that I've observed on my system).

I'd prefer a backward-compatible, silent migration; and if this breaks someone's code, I hope that they could easily adapt to the new implementation. E.g., if someone ran in an environment where the assumption of `utf8(true)` didn't hold, I would hope that they could easily adapt.

**Comment by JanWielemaker** [comment]
> with the exception of things like `re_config(utf8(X))`, where `PCRE_CONFIG_UTF8` doesn't exist in pcre2, so I'd always return `true` (the pcre1 value that I've observed on my system).

That is fine.   It was mostly a major issue if the old version was not compiled with UTF-8 support as the binding assumes it is and, without, the library will not run correctly on anything but ASCII input.   If UTF-8 is now always supported we are one step closer to reliable support.

All in all I think API changes are minimal and not that likely to affect anyone.   In the unlikely case they do, we can see what needs to be done.   That seems a lot more likely to pass without anyone noticing than including two versions.

**Comment by kamahen** [comment]
So far, I've found one change between PCRE1 and PCRE2 that will prevent backwards compatibility: in `re_config/1`, PCRE1 and PCRE2 have different integer values for `bsr` and `newline`. I propose to change these to `bsr_2` and `newline_2` and have them return atoms derived from the symbolic constants in `pcre2.h`; and throw an error if `bsr` or `newline` is used (the code already throws an "existence" error if an unknown name is used).

For the options in `re_compile/3` and `re_match/3`, I propose treating `bsr` and `bsr_2` as the same, also `newline` and `newline_2`. The atoms used by `bsr_2(V)` and `newline_2(V)` would match -- the options `bsr` and `newline` already use atoms, so there shouldn't be any need for user code changes, although PCRE2 has a few more possible values.

**Comment by JanWielemaker** [comment]
Without having looked at the code, why would different integer values for symbolic constants be of interest to the Prolog user?   I guess the integer value is not directly exposed anyway?   The value of re_config/1 is not entirely clear anyway.   It allows figuring out configuration options of the pcre library.   As on most platforms we just use the library as it is shipped with the OS/Macports/... one cannot be absolutely sure all features are present and thus if your code relies on non-standard features of the library you may wish to do a runtime test whether the library is suitable.   Possibly we should consider building the library as part of SWI-Prolog with a standard reproducible configuration so nobody has to wonder what features are present.   I typically like to keep dependencies on the outside though.

**Comment by kamahen** [comment]
The current `re_config(newline(V))` returns an integer, not an atom -- on my system, it unifies `V` with 10 (which corresponds to `lf`). Similarly for `re_config(bsr(V))`. Returning an integer for these was a mistake IMHO, so returning an atom improves the API; and it's necessary because the value corresponding. to `newline(10)` in PCRE1 is `newline(2)` in PCRE2.

**Comment by JanWielemaker** [comment]
Seems that some of the values are rightfully integers as they are character codes and others have a symbolic meaning.   Our cases are probably a bit simpler as we only use the UTF-8 mode and internal strings use canonical LF-only line breaks.   At least, if they have been read from external sources.   Users may of course put \r in a string.  If this is something we can specify at runtime we should do so.

**Comment by kamahen** [comment]
To clarify the compatible and incompatible changes for `re_config/1`:
- no change:
  - values that have intrinsic meaning will remain as numbers (e.g. `parenslimit`)
  - values that are  strings will give atom values (e.g., `version`, `jittarget`)
  - values that are 0 or 1 will be converted to `false` or `true` (e.g., `stackrecurse`)
- changes:
  - values whose values are used in an enumeration will return an atom corresponding to the pcre2.h constant (e.g., `bsr_2` (`unicode` or `anycrlf`), `newline_2` (`cr`, `lf`, `crlf`, `any`, `anycrlf`, `nul`)) and using the PCRE1 names will result in runtime exception (`bsr`, `newline`)
  - values whose names have changed in minor ways will allow both (e.g., `parens_limit` (PCRE1), `parenslimit` (PCRE2))
  - values that are no longer configuration values will either return a constant (e.g. `utf8` is always `true`) or throw an exception if they're not meaningful (e.g. `posix_malloc_threshold`)

If there's a way to configure these globally, I'll look into it -- e.g., supporting `pcre2_set_newline()`. However, that appears to require creating a "compile context" (which would presumably be a "blob" in the options to `re_compile/3` et al, so won't be in the first release.

Only the latest version's configuration values will be supported (`10.34 2019-11-21`). Versions before 10.30 had some additional configuration values that are not meaningful now.

The PCRE2 API is quite large, so there's a good chance I'll get some things wrong in the first release. Hopefully, my mistakes will be minor and easily corrected, and programs that only use PCRE1 features shouldn't be affected (new code that uses PCRE2 features might need minor changes when I fix my mistakes).

The PCRE2 API has a new `pcre2_substitute()` -- I have no plans to use this as `re_matchsub/4` should suffice. (It seems that PHP's migration to PCRE2 has made the same decision.)

**Comment by JanWielemaker** [comment]
Sounds good!

**Comment by kamahen** [comment]
One correction: the current code doesn't check for duplicate/inconsistent specification of options, so the PCRE2 code won't check for `newline`/`newline_2` and `bsr`/`bsr_2` duplicates. (Presumably this could change in future, although the documentation for `predicate_options/3` advocates for the more permissive style.)

**Comment by JanWielemaker** [comment]
Note that most of newline/bsr stuff is theory.   bsr(unicode) makes sense for applications that wish to act on arbitrary Unicode line endings.   The other newline options only make sense when working on data that has been read in binary mode, thus bypassing SWI-Prolog's input normalization for text.

**Comment by carenas** [comment]
> I think that I can preserve the current API in `pcre.pl`, with the exception of things like `re_config(utf8(X))`, where `PCRE_CONFIG_UTF8` doesn't exist in pcre2, so I'd always return `true` (the pcre1 value that I've observed on my system).

PCRE2's equivalent setting do exist and is PCRE2_CONFIG_UTF.

the default (at build time) has changed between PCRE and PCRE2 though, but most "distributions" enable unicode (which used to be disabled by default in PCRE but is enabled by default in PCRE2) by hand anyway so I would expect that unicode support will be most likely available in both, for most system provided libraries like you seem to have found on your system.

**Comment by kamahen** [comment]
I couldn't see any `PCRE2_CONFIG_UTF`. 
There are `PCRE2_UTF` (and `PCRE2_NEVER_UTF`), which can be passed to `pcre2_compile()`. 

A few of these configuration values (such as `newline`) can be set in a `pcre2_compile_context`, but I don't plan on supporting that for now (if ever).

There's also `PCRE2_CONVERT_UTF`, but that's for `pcre2_pattern_convert()` which is "experimental" and I'm not planning on supporting it.

**Comment by carenas** [comment]
> I couldn't see any `PCRE2_CONFIG_UTF`.
> There are `PCRE2_UTF` (and `PCRE2_NEVER_UTF`), which can be passed to `pcre2_compile()`.

Apologies; you are correct, I was thinking PCRE_UTF8 -> PCRE2_UTF, but should had checked first the code.

Not sure where that "define" came from, but the correct one to use if you want to check if unicode support is compiled in, as it might seem to do (after looking at the code) is PCRE2_CONFIG_UNICODE (to be passed to pcre2_config())

**Comment by JanWielemaker** [comment]
Note that Prolog's data is always UTF-8.   So all the library needs to do is to ensure pcre works in UTF-8 mode and somehow allows the user to know the core pcre library cannot do so and thus the regex library only supports US-ASCII (being the common ground between UTF-8 and plain 8-bit encoding).

**Comment by kamahen** [comment]
There PCRE2_CONFIG_UNICODE, which will show as `re_config(unicode(V))` ... I can make `re_config(utf8(V))` to be synonymous with it, to preserve how I presume people might have used PCRE1's PCRE_CONFIG_UTF8.

**Comment by kamahen** [comment]
Progress report ... I've got most things working, except for proper handling of options (5 tests that depend on option parsing plus 1 that depends on UTF8 handling fail) -- the way of setting some options seems to have changed between PCRE1 and PCRE2.
I also need to add many test cases; partly because there are a lot more options with PCRE2 and partly to check more corner cases in Unicode/UTF8 processing.
There's  good chance that things will be "good enough" by the end of the month, but it might take longer.

**Comment by JanWielemaker** [comment]
Thanks for the update.   It is nice that it also gets better!   Small question.   From a user I got a request to deal with _quoted_ regular expressions.   This of course basically kills the regex behavior but keeps the rest of the interface.   Is that something provided by PCRE2?   If not, I'd consider adding a predicate re_quote(+String, -QString) that escapes any regex special character.  PHP has something similar.

**Comment by kamahen** [comment]
Pefrhaps the "user" could open a new issue with the request for re_quote(+String, QString)? I haven't finished reading through the PCRE2 API yet ... one thing at a time. ;)

I didn't see any mention of this quoted behavior in https://wiki.php.net/rfc/pcre2-migration ... it wouldn't surprise me if it was already in the PCRE1 support.

**Comment by kamahen** [comment]
Status update -- all the tests are now succeeding, but that's because they don't test enough features (e.g., option `bsr` haas not tesst cases). I need to go through at least the existing (PCRE1) options and make sure there's an adequate test for each option; I can add PCRE2-only features later.

We can expect some things to not work with PCRE2 because it's more strict -- here's an example: https://php.watch/versions/7.3/pcre2

Here's the PR for PHP ... I hope I can get some useful tests from it:
https://github.com/php/php-src/pull/2857

**Comment by JanWielemaker** [comment]
> Pefrhaps the "user" could open a new issue with the request for re_quote(+String, QString)?

I's no secret.   It concerns @wouterbeek, working on a strict implementation of SPARQL which requires quoted regular expression matching.

**Comment by kamahen** [comment]
What is the desired functionality?
- a "raw string" (like Python's r"...") that doesn't treat backslash characters as special
- look for a substring that might contain special characters (the way fgrep does) such as "\" or "("?
Assuming the latter, why not use sub_string/5 ?

**Comment by wouterbeek** [comment]
Hi @kamahen , others.  I'm interested in the functionality of the "q" flag as defined in the XPath standard: <https://www.w3.org/TR/xpath-functions/#flags>

The others flags are already covered by the SWI implementation I think.  The flags in XPath standard are reused in SPARQL.

**Comment by JanWielemaker** [comment]
> * Assuming the latter, why not use sub_string/5

Good question.  I don't know how that compares to @wouterbeek's problem.   It it is about matching only, it is enough (I guess).  If you want to use the more complicated operations such as splitting it is nice to be able to reuse the regex APIs.   I've found a possibility to quote a string such that it is taken literally useful when composing a regular expression from a string that comes from "outside".   E.g. "given an arbitrary string, find all matches at the start of the line" can be achieved by quoting the string and adding `^`  at the start.    Another example could be case insensitive matching for an arbitrary string.

**Comment by kamahen** [comment]
Hi @wouterbeek -- It appears that the `\Q` ... `\E` sequence will do what you want: https://www.pcre.org/current/doc/html/pcre2pattern.html#SEC5 

Therefore, it would seem that we could define
```
re_quote(String, Qstring) : -
    atomics_to_string(['\\Q', String, '\\E'], Qstring).
```
Presumably this will behave correctly with Unicode whose UTF8 happens to contain characters such as "(" and "\\" ... not sure how to easily find test cases for that, though.

**Comment by wouterbeek** [comment]
@kamahen Thanks!  `\Q...\E` seems to be working as an implementation of the XPath `x` flag.

**Comment by kamahen** [comment]
`\Q...\E` might not be sufficient: "The \Q...\E sequence is recognized both inside and outside character classes."
So, some additional processing might be needed for `[...]`, although I'm having trouble seeing how `\Q...\E` might be used in a character class (perhaps there are some test cases at https://github.com/PhilipHazel/pcre2).

**Comment by wouterbeek** [comment]
@kamahen Thanks for the relevant warning.  I'm going to try out `\Q...\E` and will report any issues that we find back here.

**Comment by kamahen** [comment]
You might find the `pcre2grep` command useful -- it's in Ubuntu package `pcre2-utils`, which also has `pcre2test`.

**Comment by kamahen** [comment]
`README.mingw` has the line `wget https://ftp.pcre.org/pub/pcre/$PCRE_FILE` ... this doesn't work (https with ftp); and the sources seem to now be entirely on github (the ftp site doesn't seem to exist) [also, for PCRE1, PCRE_VERSION should have been updated from 8.43 to 8.45]

I have no ability to test mingw, so what do you suggest doing (if anything)?

**Comment by kamahen** [comment]
`scripts/macos.sh` has a similar problem to mingw.

**Comment by JanWielemaker** [comment]
> I have no ability to test mingw, so what do you suggest doing (if anything)?

For now, nothing as far as I'm concerned.   The Windows build gets PCRE from the `mingw32-pcre` from Fedora.   We'll move to the latest after all compiles against pcre2.

**Comment by kamahen** [comment]
For the record, here are all the places that need to be changed outside of `packages/pcre` (there might be more, of course):
```diff
diff --git a/.travis.yml b/.travis.yml
index fe1c7642f..f7a983c81 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -32,7 +32,7 @@ addons:
       - libxpm-dev
       - libxt-dev
       - libdb-dev
-      - libpcre3-dev
+      - libpcre2-dev
       - libyaml-dev
       - openjdk-8-jdk
       - junit
@@ -49,7 +49,7 @@ addons:
       - libyaml
       - unixodbc
       - berkeley-db
-      - pcre
+      - pcre2
       - jpeg
 
 # The  code  below  may  be  used  to   work  from  your  own  clone  of
```
```diff
diff --git a/CMAKE.md b/CMAKE.md
index 86dc6fee4..51280f180 100644
--- a/CMAKE.md
+++ b/CMAKE.md
@@ -389,7 +389,7 @@ The defined components are:
   | Commandline_editors  | Readline and libedit interfaces      |
   | ODBC_interface       | ODBC binding                         |
   | BerkeleyDB_interface | BDB interface                        |
-  | Perl_regex           | PCRE library binding                 |
+  | Perl_regex           | PCRE2 library binding                |
   | YAML_support         | Libyaml binding                      |
   | Java_interface       | Java interface (JPL)                 |
   | OpenSSL_interface    | Binding to OpenSSL/LibreSSL          |
```
```diff
diff --git a/README.mingw b/README.mingw
index 150c514bd..67c07e57e 100644
--- a/README.mingw
+++ b/README.mingw
@@ -123,7 +123,7 @@ ZLIB_VERSION=1.2.11
 ARCHIVE_VERSION=3.4.0
 UUID_VERSION=1.6.2
 BDB_VERSION=6.1.26
-PCRE_VERSION=8.43
+PCRE_VERSION=8.45  # TODO: update for PCRE2
 FFI_VERSION=3.3
 YAML_VERSION=0.2.2
 PLDIR=pl-devel
@@ -455,6 +455,7 @@ build_libarchive()
 # Download and install libpcre
 
 
+# TODO: update for PCRE2
 download_libpcre()
 { PCRE_FILE=pcre-$PCRE_VERSION.tar.gz
 
```
```diff
diff --git a/cmake/cross/linux_i386.cmake b/cmake/cross/linux_i386.cmake
index 8c3c2ff88..43e588065 100644
--- a/cmake/cross/linux_i386.cmake
+++ b/cmake/cross/linux_i386.cmake
@@ -14,7 +14,7 @@
 #	libdb-dev:i386 libedit-dev:i386 libgmp-dev:i386 \
 #	libice-dev:i386 libjpeg-dev:i386 libjpeg8-dev:i386 \
 #	libncurses5-dev:i386 libossp-uuid-dev:i386 \
-#	libpcre3-dev:i386 libssl1.0-dev:i386 \
+#	libpcre2-dev:i386 libssl1.0-dev:i386 \
 #	libstdc++-7-dev:i386 libxext-dev:i386 libxft-dev:i386 \
 #	libxinerama-dev:i386 libxpm-dev:i386 libxt-dev:i386 \
 #	libyaml-dev:i386 zlib1g-dev:i386
```
```diff
diff --git a/scripts/macos-deps.sh b/scripts/macos-deps.sh
index 1d9e7cc23..89d567b80 100644
--- a/scripts/macos-deps.sh
+++ b/scripts/macos-deps.sh
@@ -23,7 +23,7 @@ ARCHIVE_VERSION=3.5.2
 UUID_VERSION=1.6.2
 BDB_VERSION=6.1.26
 ODBC_VERSION=2.3.9
-PCRE_VERSION=8.43
+PCRE_VERSION=8.45  # TODO: update for PCRE2
 FFI_VERSION=3.4.2
 YAML_VERSION=0.1.7
 READLINE_VERSION=8.0
@@ -223,6 +223,7 @@ build_libarchive()
 # Download and install libpcre
 
 
+# TODO: update for PCRE2
 download_libpcre()
 { PCRE_FILE=pcre-$PCRE_VERSION.tar.gz
 
```
```diff
diff --git a/snap/snapcraft.yaml b/snap/snapcraft.yaml
index 97d2650d7..5d7640ff9 100644
--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -84,7 +84,7 @@ parts:
       # - libssl1.1
       # - libx11-6
       - libossp-uuid16
-      - libpcre3
+      - libpcre2
       - libserd-0-0
       - libxft2
       - libxpm4
@@ -102,7 +102,7 @@ parts:
       - libice-dev
       - libjpeg-dev
       - libossp-uuid-dev
-      - libpcre3-dev
+      - libpcre2-dev
       - libraptor2-dev
       - libreadline-dev
       - libserd-dev
```
In addition, the following change to `cmake/BuildType.cmake` has been useful to find some potential problems with the code:
```
set(CMAKE_C_FLAGS_RELWITHDEBINFO "-O2 -gdwarf-2 -g3 -Wextra -Wuninitialized -Wmaybe-uninitialized -Wno-missing-field-initializers"
```

**Comment by kamahen** [comment]
I've integrated all the PCRE1-related changes (up to https://github.com/SWI-Prolog/packages-pcre/commit/eda1472a69976de602d390a72ec661ecd71cfebf) with the pcre2 changes, and most the tests pass. I'm now working on the options that are unique to PCRE2. Also, some of my tests need to be modified because there's no obvious way to extract the `newlines` and `bsr` flags, so I need to figure out a different way of testing those.

**Comment by JanWielemaker** [comment]
Thanks for the diff.  That will come handy.   I suppose otherwise I should simply wait?

**Comment by kamahen** [comment]
My plan is to create a git patch, which I presume will be easier for you to handle than a diff.
(It's somewhat more work for me to create PRs, so I'd prefer creating the git patch)

**Comment by kamahen** [comment]
I've pushed an initial version of the PCRE2 changes in PR https://github.com/SWI-Prolog/packages-pcre/pull/11

This code needs some review, to ensure that all the options are handled properly (a couple of tests don't pass right now, probably due to typos).
If anyone wants to help with this review, please coordinate with me, so that we don't duplicate each others' work.

**Comment by kamahen** [comment]
After further reading, it seems that I do *not* need any mutex for `pcre2_jit_optimize()` if it's called right after `pcre2_compile()` (which is what I intend to do, replacing the call to `pcre_study()` which has no equivalent in pcre2), and there's no need for a global mutex. 
There are some more options that are specific to `pcre2_jit_optimize()` - I'll open a new issue, as I go through the details.

**Comment by JanWielemaker** [comment]
SWI-Prolog 8.5.9 now uses pcre2.    Thanks @kamahen !    Considering the vast amount of changes and small incompatibilities this will probably not be backported to 8.4.x.
